File: apis/elysium-chat-api/src/main/java/com/seventh_root/elysium/api/chat/ChatChannelProvider.java
package com.seventh_root.elysium.api.chat;

import com.seventh_root.elysium.api.player.ElysiumPlayer;
import com.seventh_root.elysium.core.service.ServiceProvider;

import java.util.Collection;

public interface ChatChannelProvider<T extends ElysiumChatChannel> extends ServiceProvider {

    Collection<? extends T> getChatChannels();

    T getChatChannel(int id);

    T getChatChannel(String name);

    void addChatChannel(T chatChannel);

    void removeChatChannel(T chatChannel);

    void updateChatChannel(T chatChannel);

    T getPlayerChannel(ElysiumPlayer player);

    void setPlayerChannel(ElysiumPlayer player, T channel);

    T getChatChannelFromIRCChannel(String ircChannel);

}

File: bukkit/elysium-chat-bukkit/src/main/kotlin/com/seventh_root/elysium/chat/bukkit/chatchannel/ChatChannelProvider.kt
package com.seventh_root.elysium.chat.bukkit.chatchannel

import com.seventh_root.elysium.chat.bukkit.chatchannel.ElysiumChatChannel
import com.seventh_root.elysium.players.bukkit.player.ElysiumPlayer
import com.seventh_root.elysium.core.service.ServiceProvider

interface ChatChannelProvider<T : ElysiumChatChannel> : ServiceProvider {

    val chatChannels: Collection<T>
    fun getChatChannel(id: Int): T?
    fun getChatChannel(name: String): T?
    fun addChatChannel(chatChannel: T)
    fun removeChatChannel(chatChannel: T)
    fun updateChatChannel(chatChannel: T)
    fun getPlayerChannel(player: ElysiumPlayer): T?
    fun setPlayerChannel(player: ElysiumPlayer, channel: T)
    fun getChatChannelFromIRCChannel(ircChannel: String): T?

}
